tesseract/consensus/polygon-pos/src/lib.rs

```diff
@@ -1,122 +1,183 @@
+		Ok(Arc::new(PolygonPosHost::new(&self.host, &self.evm_config).await?))
 	}
 
 	pub fn state_machine(&self) -> StateMachine {
 		self.evm_config.state_machine
 	}
 }
 
+/// The relayer host for Polygon POS
 #[derive(Clone)]
 pub struct PolygonPosHost {
-	/// Consensus state id on counterparty chain
 	pub consensus_state_id: ConsensusStateId,
-	/// State machine Identifier for this chain.
 	pub state_machine: StateMachine,
-	/// Consensus prover
-	pub prover: PolygonPosProver,
-	/// Config
-	pub config: PolygonPosConfig,
-	/// Jsonrpsee client for event susbscription, ethers does not expose a Send and Sync stream for
-	/// susbcribing to contract logs
-	pub rpc_client: Arc<Client>,
+	pub host: HostConfig,
+	pub evm: EvmConfig,
+	pub provider: Arc<dyn IsmpProvider>,
+	pub prover: HeimdallClient,
 }
 
 impl PolygonPosHost {
-	pub async fn new(config: &PolygonPosConfig) -> Result<Self, anyhow::Error> {
-		let provider =
-			Provider::<Http>::try_from(config.evm_config.rpc_url.clone())?;
-		let prover = PolygonPosProver::new(provider);
-		let rpc_client = Client::builder()
-			.retry_policy(ExponentialBackoff::from_millis(100))
-			.enable_ws_ping(
-				PingConfig::new()
-					.ping_interval(Duration::from_secs(6))
-					.inactive_limit(Duration::from_secs(30)),
-			)
-			.build(config.evm_config.rpc_url.clone())
-			.await?;
+	/// Create a new PolygonPosHost
+	pub async fn new(host: &HostConfig, evm: &EvmConfig) -> Result<Self, anyhow::Error> {
+		let ismp_provider = EvmClient::new(evm.clone()).await?;
 		Ok(Self {
 			consensus_state_id: {
 				let mut consensus_state_id: ConsensusStateId = Default::default();
-				consensus_state_id.copy_from_slice(config.evm_config.consensus_state_id.as_bytes());
+				consensus_state_id.copy_from_slice(evm.consensus_state_id.as_bytes());
 				consensus_state_id
 			},
-			state_machine: config.evm_config.state_machine,
-			prover,
-			config: config.clone(),
-			rpc_client: Arc::new(rpc_client),
+			state_machine: evm.state_machine,
+			host: host.clone(),
+			evm: evm.clone(),
+			provider: Arc::new(ismp_provider),
+			prover: HeimdallClient::new(
+				&host.heimdall_rpc_url,
+				&host.heimdall_rest_url,
+				&host.execution_rpc_url,
+			)?,
 		})
 	}
 
-	pub async fn get_consensus_state(
-		&self,
-		ismp_contract_address: H160,
-	) -> Result<ConsensusState, anyhow::Error> {
-		let (header, validators) = self.prover.fetch_finalized_state().await?;
-
-		let consensus_state = ConsensusState {
-			frozen_height: None,
-			finalized_hash: Header::from(&header).hash::<KeccakHasher>(),
-			finalized_validators: validators,
-			forks: Default::default(),
-			ismp_contract_address,
-		};
+	/// Fetch the current consensus state (for initial state creation)
+	pub async fn get_consensus_state(&self) -> Result<ConsensusState, anyhow::Error> {
+		let consensus_state_serialized: Vec<u8> =
+			self.provider.query_consensus_state(None, self.consensus_state_id).await?;
```

Comment:
This is wrong, don't use the provider when getting initial consensus state, construct it.